Задание 3 (Пул потоков)

Cherkashin-AV opened the pull request.

Cherkashin-AV committed:
first commit

Alexandr056 commented on src/main/java/ru/vtb/jpro/mthread/ThreadPool.java:
```diff
@@ -0,0 +1,76 @@
+    private void runNextTaskFromQueue() {
+        Runnable curTask;
+        while (true) {
+            synchronized(this) {
```
я бы синхронизировался не по всему пулу потоков, можно queue взять

Cherkashin-AV replied:
Исправлено

Alexandr056 commented on src/main/java/ru/vtb/jpro/mthread/ThreadPool.java:
```diff
@@ -0,0 +1,76 @@
+            synchronized(this) {
+                while (queue.isEmpty()) {
+                    try {
+                        Thread.sleep(timeout);
```
не думаю что sleep это хорошее решение, мы говорили на уроке про тайм ауты и почему это не очень оптимально, почем придуман механизм wait notify()
здесь стоит переводить поток в состояние wait

Cherkashin-AV replied:
Исправлено

Alexandr056 commented on src/main/java/ru/vtb/jpro/mthread/ThreadPool.java:
```diff
@@ -0,0 +1,76 @@
+            throw new  IllegalStateException("Пул потоков в процессе остановки");
+        }
+        synchronized(queue) {
+            queue.addLast(task);
```
а здесь добавить notifyAll так как задача поступила можно разбудить поток для выполнения

Cherkashin-AV replied:
Исправлено

Alexandr056 commented on src/main/java/ru/vtb/jpro/mthread/ThreadPool.java:
```diff
@@ -0,0 +1,76 @@
+        isShutDown = true;
+    }
+
+    public void awaitTermination() throws InterruptedException {
```
Метод awaitTermination у ExecutorService позволяет безопасно завершать работу пула потоков, обеспечивая возможность ожидания завершения всех задач в указанное время.

по сути как join внутри пула, синхронизация по чему-либо (рассмотрели на втором уроке)
но здесь просто засыпание?

Cherkashin-AV replied:
Исправлено. Переделал на join всех рабочих потоков

Alexandr056 reviewed (changes requested):
немного корректировок по wait notify и будет отлично

Cherkashin-AV committed:
Устранение замечаний

1. Ожидание поступления задачи переделано на wait/notify
2. awaitTermination переделан на join всех потоков

Alexandr056 commented on src/main/java/ru/vtb/jpro/mthread/ThreadPool.java:
```diff
@@ -0,0 +1,79 @@
+        //Добавляем задачу в очередь и оповещяем спящий поток, что пришла задача
+        synchronized(queue) {
+            queue.addLast(task);
+            queue.notify();
```
notifyAll() мы же не знаем сколько там потоков, пробуждаем все, а они уже кто найдет себе работу тот найдет

Cherkashin-AV replied:
Потоки выполняют одну и туже работу, а к нам пришел один запрос. Зачем будить все потоки, когда нам нужен всего один произвольный поток для обработки одного запроса?

Alexandr056 replied:
хотя может и так, я представил что задачи могут пачкой добавляться и речь не про один поток
в целом да можно и так

разница в том, что notify() может быть более эффективным, если известно, что только один поток ожидает доступа к монитору. В противном случае, если несколько потоков ожидают доступа, и необходимо гарантировать, что все они будут уведомлены, когда ресурс станет доступным, следует использовать notifyAll().

Alexandr056 commented on src/main/java/ru/vtb/jpro/mthread/ThreadPool.java:
```diff
@@ -0,0 +1,79 @@
+    private void runNextTaskFromQueue() {
+        Runnable curTask;
+        while (true) {
+            synchronized(queue) {
```
многова-то кода в блоке синхронизации, зесь по сути нужна синхронизация только для обращения к коллекции

Cherkashin-AV replied:
Здесь еще notifyAll и wait, которые должны выполняться в блоке synchronized с обработкой ошибок. С моей точки зрения тут вынести из блока synchronized  нечего.

Alexandr056 reviewed (approved):
работу принимаю, есть ряд замечаний, но это часто понятно из-за сложности понимания многопоточности, спасибо за проделанную работу
